lib/spack/spack/solver/asp.py

```diff
@@ -1394,10 +1394,6 @@ def define_version_constraints(self):
             if exact_match:
                 allowed_versions = exact_match
 
-            # don't bother restricting anything if all versions are allowed
-            if len(allowed_versions) == len(self.possible_versions[pkg_name]):
-                continue
-
```

Comment:
I think this is the right modification to solve the issue with `when-directives-true`.